File: ljh/level2/거리두기 확인하기/DistancingCheck.js
// 거리두기 확인하기

function solution(places) {
    const answer = [];
    places.forEach((v) => answer.push(CheckDistaning(v)? 1 : 0));
    return answer;
}

function CheckDistaning(arr) { 
    const searchP = Array.from(Array(5), () => Array(5).fill(0));    
    for (let i = 0; i < 5; i++) {
        for (let j = 0; j < 5; j++) {
            if(arr[i][j] === 'P') {
                searchP[i][j] += 1;
                if(i - 1 >= 0) 
                    searchP[i-1][j] += 1;
                if(i + 1 < 5) 
                    searchP[i+1][j] += 1;
                if(j - 1 >= 0) 
                    searchP[i][j-1] += 1;
                if(j + 1 < 5) 
                    searchP[i][j+1] += 1;
            }
        }
    }
    
    for (let i = 0; i < 5; i++) {
        for (let j = 0; j < 5; j++) {
            if(searchP[i][j] > 1 && arr[i][j] != 'X') {
                return false;
            }
        }
    }
    
    return true;
}

File: ljh/level2/거리두기 확인하기/DistancingCheck.java
package Programmers.test;

/**
 * 거리두기 확인하기
 **/

public class DistancingCheck {
    public int[] solution(String[][] places) {
        int[] answer = new int[5];
        for (int i = 0; i < places.length; i++) {
            answer[i] = distancing(places[i]);
        }
        return answer;
    }

    public int distancing(String[] arr) {
        for (int i = 0; i < arr.length; i++) {
            for (int j = 0; j < 5; j++) {
                if (arr[i].charAt(j) == 'P' && !move(arr, j, i, 0, 0))
                    return 0;
            }
        }
        return 1;
    }

    public boolean move(String[] arr, int x, int y, int n, int next) { // -1 : left, 0 : down, 1 : right
        if (n > 2 || y == 5 || x == -1 || x == 5)
            return true;

        char c = arr[y].charAt(x);
        if (n != 0 && c == 'P')
            return false;
        if (c == 'X')
            return true;
        if (next == 0)
            return move(arr, x + 1, y, n + 1, 1) && move(arr, x -1, y, n + 1, -1) && move(arr, x, y + 1, n + 1, 0);

        return move(arr, x + next, y, n + 1, next) && move(arr, x, y + 1, n + 1, next);
    }
}
